libcst/codegen/gen_visitor_functions.py

```diff
@@ -0,0 +1,163 @@
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+# pyre-strict
+import inspect
+from collections import defaultdict
+from typing import Dict, Generator, List, Mapping, Sequence, Set, Type
+
+import libcst as cst
+
+
+DO_NOT_CARE_TYPES: Set[str] = {"CSTNode", "ABC", "object"}
+
+
+def _get_nodes() -> Generator[Type[cst.CSTNode], None, None]:
+    """
+    Grab all CSTNodes that are not a superclass. Basically, anything that a
+    person might use to generate a tree.
+    """
+
+    # pyre-ignore __dir__ exists on all objects.
+    for name in cst.__dir__():
+        if name.startswith("__") and name.endswith("__"):
+            continue
+        if name == "CSTNode":
+            continue
+
+        node = getattr(cst, name)
+        try:
+            if issubclass(node, cst.CSTNode):
+                yield node
+        except TypeError:
+            # This isn't a class, so we don't care about it.
+            pass
+
+
+nodes: Sequence[Type[cst.CSTNode]] = list(_get_nodes())
+classmap: Dict[Type[cst.CSTNode], Set[Type[cst.CSTNode]]] = {}
+for node in nodes:
+    # Map the base classes for this node
+    classmap[node] = set(inspect.getmro(node))
+
+
+def _get_bases(node: Type[cst.CSTNode]) -> Generator[Type[cst.CSTNode], None, None]:
+    for base in classmap[node]:
+        if base.__name__ in DO_NOT_CARE_TYPES:
+            # Ignore bases that are Python implementation details.
+            continue
+        if base not in classmap:
+            # Ignore bases that aren't in our tracking list.
+            continue
+        if base.__name__.startswith("_Base"):
+            # Ignore bases that aren't exported.
+            continue
+        yield base
+
+
+def _get_base(node: Type[cst.CSTNode]) -> Type[cst.CSTNode]:
+    bases = set(_get_bases(node))
+    if not bases:
+        raise Exception(f"Couldn't get any bases for {node.__name__}!")
+    if len(bases) == 1:
+        for b in bases:
+            return b
+    bases = {b for b in bases if b is not node}
+
+    new_bases = set()
+    for base in bases:
+        new_bases.add(_get_base(base))
+
+    if len(new_bases) > 1:
+        raise Exception(f"Got too many bases for {node.__name__}!")
+    if len(new_bases) == 1:
+        for b in new_bases:
+            return b
+    raise Exception(f"Couldn't get any bases for {node.__name__}!")
+
+
+nodebases: Dict[Type[cst.CSTNode], Type[cst.CSTNode]] = {}
```

Comment:
Between `nodes`, `classmap`, and `nodebases` it can be hard to figure out what variable does what. I think it would help if you gave these variables more detailed names.